scripts/sandbox.js

```diff
@@ -1,40 +1,66 @@
-let sandboxLevel = "";
-let sandboxObject = ObstacleCube;
-let sanboxTpX = 3;
-let sandboxTpY = 3;
-let sandboxPortalColor = "Blue";
+const sandbox = {
+  history: [null],
+  level: "",
+  object: ObstacleCube,
+  tpX: 0,
+  tpY: 0,
+  portalColor: "Blue",
+};
 
 class Sandbox {
   static update() {
     UserInterface.displayText({
-      content: `x: ${Math.floor(mouseX / UNIT_SIZE)}, y:${Math.floor(
+      content: `x: ${Math.floor(mouseX / UNIT_SIZE)}, y: ${Math.floor(
         mouseY / UNIT_SIZE
       )}`,
       x: 1,
       y: 2,
     });
     this.drawMarking(mouseX, mouseY);
   }
+
+  static undo() {
+    const history = sandbox.history;
+
+    if (history.length - 1 > 0) {
+      history.splice(history.length - 1, 1);
+    }
+  }
+
   static pushLevel() {
+    const history = sandbox.history;
+
     scenes.splice(scenes.length - 1, 1, () => {
-      eval(sandboxLevel);
+      sandbox.level = history[history.length - 1];
+      eval(sandbox.level);
```

Comment:
FYI. `eval` has security implications. You don't **have** to change this, but I think a cleaner and safer way would be to instead save an array of arrays, where each array represents what should be drawn.

For example:

```javascript
[
  [
    { type: "Obstacle", x: 1, y: 1, portalColor: "red", tpX: 1, tpY: 1 },
    { type: "Obstacle", x: 1, y: 1, portalColor: "red", tpX: 1, tpY: 1 },
  ]
]
```

Reply:
I agree, I was just lazy in the beginning of trying to fix our sandbox mode. I was thinking about doing something like this to instead be able to delete cubes by clicking at them.